admittance_controller/include/admittance_controller/admittance_rule.hpp

```diff
@@ -0,0 +1,215 @@
+#include "geometry_msgs/msg/wrench_stamped.hpp"
+#include "kinematics_interface/kinematics_interface.hpp"
+#include "pluginlib/class_loader.hpp"
+#include "tf2_eigen/tf2_eigen.hpp"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include "tf2_kdl/tf2_kdl/tf2_kdl.hpp"
+#include "tf2_ros/buffer.h"
+#include "tf2_ros/transform_listener.h"
+#include "trajectory_msgs/msg/joint_trajectory_point.hpp"
+
+namespace admittance_controller
+{
+struct AdmittanceTransforms
+{
+  Eigen::Isometry3d ref_base_ft_;
+  Eigen::Isometry3d base_ft_;
+  Eigen::Isometry3d base_control_;
+  Eigen::Isometry3d base_tip_;
+  Eigen::Isometry3d world_base_;
+  Eigen::Isometry3d base_sensor_;
+  Eigen::Isometry3d base_cog_;
+};
+
+struct AdmittanceState
+{
+  explicit AdmittanceState(size_t num_joints)
+  {
+    admittance_velocity.setZero();
+    admittance_acceleration.setZero();
+    damping.setZero();
+    mass.setOnes();
+    mass_inv.setZero();
+    stiffness.setZero();
+    selected_axes.setZero();
+    current_joint_pos = Eigen::VectorXd::Zero(num_joints);
+    joint_pos = Eigen::VectorXd::Zero(num_joints);
+    joint_vel = Eigen::VectorXd::Zero(num_joints);
+    joint_acc = Eigen::VectorXd::Zero(num_joints);
+  }
+
+  Eigen::VectorXd current_joint_pos;
+  Eigen::VectorXd joint_pos;
+  Eigen::VectorXd joint_vel;
+  Eigen::VectorXd joint_acc;
+  Eigen::Matrix<double, 6, 1> damping;
+  Eigen::Matrix<double, 6, 1> mass;
+  Eigen::Matrix<double, 6, 1> mass_inv;
+  Eigen::Matrix<double, 6, 1> selected_axes;
+  Eigen::Matrix<double, 6, 1> stiffness;
+  Eigen::Matrix<double, 6, 1> wrench_base;
+  Eigen::Matrix<double, 6, 1> admittance_acceleration;
+  Eigen::Matrix<double, 6, 1> admittance_velocity;
+  Eigen::Isometry3d admittance_position;
+  Eigen::Matrix<double, 3, 3> rot_base_control;
+  Eigen::Isometry3d ref_trans_base_ft;
+  std::string ft_sensor_frame;
+};
+
+class AdmittanceRule
+{
+public:
+  explicit AdmittanceRule(
+    const std::shared_ptr<admittance_controller::ParamListener> & parameter_handler)
+  {
+    parameter_handler_ = parameter_handler;
+    parameters_ = parameter_handler_->get_params();
+    num_joints_ = parameters_.joints.size();
+    admittance_state_ = AdmittanceState(num_joints_);
+    reset(num_joints_);
+  }
+
+  /// Configure admittance rule memory using number of joints.
+  controller_interface::return_type configure(
+    const std::shared_ptr<rclcpp_lifecycle::LifecycleNode> & node, const size_t num_joint);
+
+  /// Reset all values back to default
+  controller_interface::return_type reset(const size_t num_joints);
+
+  /**
+   * Calculate all transforms needed for admittance control using the loader kinematics plugin. If the transform does
+   * not exist in the kinematics model, then TF will be used for lookup. The return value is true if all transformation
+   * are calculated without an error
+   * \param[in] current_joint_state
+   * \param[in] reference_joint_state
+   * \param[out] success
```

Comment:
Why not to add some description here to parameters? I understand that is “stating the obvious”, but it is also good to have some more information in the long term. Just write as you have to explain them to someone.

Reply:
done already